File: src/integration.f90
module integration
    use kinds, only: dp
    use errors, only: require
    implicit none

    abstract interface
        function scalar_fun(x) result(fx)
            import dp
            real(dp), intent(in) :: x
            real(dp) :: fx
        end function scalar_fun
    end interface
contains
    function trapezoidal(f, a, b, n) result(area)
        procedure(scalar_fun) :: f
        real(dp), intent(in) :: a, b
        integer, intent(in) :: n
        real(dp) :: area
        real(dp) :: h
        integer :: i
        call require(n >= 1, 'trapezoidal: n must be >= 1')
        h = (b - a) / real(n, dp)
        area = 0.5_dp * (f(a) + f(b))
        do i = 1, n-1
            area = area + f(a + h * real(i, dp))
        end do
        area = area * h
    end function trapezoidal

    function simpson(f, a, b, n) result(area)
        procedure(scalar_fun) :: f
        real(dp), intent(in) :: a, b
        integer, intent(in) :: n
        real(dp) :: area
        real(dp) :: h
        integer :: i
        call require(mod(n,2) == 0, 'simpson: n must be even')
        h = (b - a) / real(n, dp)
        area = f(a) + f(b)
        do i = 1, n-1
            if (mod(i,2) == 0) then
                area = area + 2.0_dp * f(a + h * real(i, dp))
            else
                area = area + 4.0_dp * f(a + h * real(i, dp))
            end if
        end do
        area = area * h / 3.0_dp
    end function simpson

    function adaptive_trapezoidal(f, a, b, tol, max_iter, iterations, converged) result(area)
        procedure(scalar_fun) :: f
        real(dp), intent(in) :: a, b, tol
        integer, intent(in) :: max_iter
        integer, intent(out) :: iterations
        logical, intent(out) :: converged
        real(dp) :: area
        real(dp) :: prev_area, curr_area
        integer :: n

        call require(tol > 0.0_dp, 'adaptive_trapezoidal: tol must be > 0')
        n = 1
        prev_area = trapezoidal(f, a, b, n)
        converged = .false.
        do iterations = 1, max_iter
            n = n * 2
            curr_area = trapezoidal(f, a, b, n)
            if (abs(curr_area - prev_area) < tol) then
                converged = .true.
                area = curr_area
                return
            end if
            prev_area = curr_area
        end do
        area = curr_area
    end function adaptive_trapezoidal
end module integration

File: python_solver/integration.py
from __future__ import annotations

from typing import Callable, Tuple


def trapezoidal(f: Callable[[float], float], a: float, b: float, n: int) -> float:
    h = (b - a) / n
    area = 0.5 * (f(a) + f(b))
    for i in range(1, n):
        area += f(a + h * i)
    return area * h


def simpson(f: Callable[[float], float], a: float, b: float, n: int) -> float:
    if n % 2 != 0:
        raise ValueError("n must be even")
    h = (b - a) / n
    area = f(a) + f(b)
    for i in range(1, n):
        coeff = 4.0 if i % 2 == 1 else 2.0
        area += coeff * f(a + h * i)
    return area * h / 3.0


def adaptive_trapezoidal(
    f: Callable[[float], float],
    a: float,
    b: float,
    tol: float,
    max_iter: int,
) -> Tuple[float, int, bool]:
    n = 1
    prev = trapezoidal(f, a, b, n)
    for iteration in range(1, max_iter + 1):
        n *= 2
        curr = trapezoidal(f, a, b, n)
        if abs(curr - prev) < tol:
            return curr, iteration, True
        prev = curr
    return curr, max_iter, False
